searches/search_test.go

```diff
@@ -1,49 +1,65 @@
 package searches
 
-import "testing"
+import (
+	"testing"
+)
 
 type searchTest struct {
-	data     []int
-	key      int
-	expected int
-	name     string
+	data          []int
+	key           int
+	expectedValue int
+	expectedError error
+	name          string
 }
 
 var searchTests = []searchTest{
 	//Sanity
-	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 1, 0, "Sanity"},
-	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 5, 4, "Sanity"},
-	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 10, 9, "Sanity"},
+	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 10, 9, nil, "Sanity"},
+	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 9, 8, nil, "Sanity"},
+	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 8, 7, nil, "Sanity"},
+	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 7, 6, nil, "Sanity"},
+	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 6, 5, nil, "Sanity"},
+	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 5, 4, nil, "Sanity"},
+	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 4, 3, nil, "Sanity"},
+	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 3, 2, nil, "Sanity"},
+	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 2, 1, nil, "Sanity"},
+	{[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 1, 0, nil, "Sanity"},
 	//Absent
-	{[]int{1, 4, 5, 6, 7, 10}, -25, -1, "Absent"},
-	{[]int{1, 4, 5, 6, 7, 10}, 25, -1, "Absent"},
+	{[]int{1, 4, 5, 6, 7, 10}, -25, -1, errNotFound, "Absent"},
+	{[]int{1, 4, 5, 6, 7, 10}, 25, -1, errNotFound, "Absent"},
 	//Empty slice
-	{[]int{}, 2, -1, "Empty"},
+	{[]int{}, 2, -1, errNotFound, "Empty"},
 }
 
 func TestBinarySearch(t *testing.T) {
 	for _, test := range searchTests {
-		actual := BinarySearch(test.data, test.key, 0, len(test.data)-1)
-		if actual != test.expected {
-			t.Errorf("test %s failed", test.name)
+		actualValue, actualError := BinarySearch(test.data, test.key, 0, len(test.data)-1)
+		if actualValue != test.expectedValue {
+			t.Errorf("test '%s' failed: input array '%v' with key '%d', expected '%d', get '%d'", test.name, test.data, test.key, test.expectedValue, actualValue)
+		}
+		if actualError != test.expectedError {
+			t.Errorf("test '%s' failed: input array '%v' with key '%d', expected error '%s', get error '%s'", test.name, test.data, test.key, test.expectedError, actualError)
 		}
 	}
 }
 
```

Comment:
it should be something along the lines of

```suggestion

func BenchmarkBinarySearch(b *testing.B) {
	var testCase []int
	for i := 0; i < b.N; i++ {
		testCase = append(testCase, i)
	} // You can take this completely outside with a reasonably large size say 100 - that way it is easier to take this and use it inside other benchmarks too for comparison of efficiency and speed.
	b.ResetTimer() // this is important because the above loop is expensive
	for i := 0; i < b.N; i++ {
		_, _ := BinarySearch(testCase, 10, 0, len(testCase) - 1)
	}
}

```

In the comment I meant this

```suggestion

func generateBenchmarkTestCase() []int {
	var testCase []int
	for i := 0; i < 100; i++ {
		testCase = append(testCase, i)
	}
	return testCase
}

func BenchmarkBinarySearch(b *testing.B) {
	testCase := generateBenchmarkTestCase()
	b.ResetTimer() // this is important because the generateBenchmarkTestCase() is expensive
	for i := 0; i < b.N; i++ {
		_, _ := BinarySearch(testCase, 10, 0, len(testCase) - 1)
	}
}

```

Make sure to run `gofmt -s` on the changed files for any formatting issue.

You should read up on the [testing](https://golang.org/pkg/testing/) documentation. It's quite good and very (very) useful.

In the go standard sort library that you linked, you should look at [this](https://github.com/golang/go/blob/618b66e16d5adedbeccb7e7ed35b26e439a9f86f/src/sort/search_test.go#L143) line, this is where they test the actually benchmark the sorting algorithms on in-built types.

Note that the efficiency testing that they are doing is a `test` not a `benchmark` thus it is testing something not benchmarking and you can't use cpu profilers on them (easily).

Reply:
Thanks for the guidance!